src/vdo.rs

```diff
@@ -0,0 +1,482 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: © 2024 Google
+// Ported from libtypec (Rajaram Regupathy <rajaram.regupathy@gmail.com>)
+
+//! The VDO data structures
+
+#[cfg(feature = "backtrace")]
+use std::backtrace::Backtrace;
+use std::ffi::CString;
+use std::os::unix::ffi::OsStrExt;
+
+use bitstream_io::BitRead;
+use enumn::N;
+use proc_macros::Printf;
+use proc_macros::Snprintf;
+
+use crate::BcdWrapper;
+use crate::BitReader;
+use crate::Error;
+use crate::FromBytes;
+use crate::Result;
+
+#[repr(C)]
+/// Maximum VPD VBUS Voltage
+#[derive(Debug, Clone, Copy, PartialEq, Eq, N)]
+pub enum Pd3p2MaxVbusVoltage {
+    /// 20V
+    V20 = 0,
+    /// 30V (Deprecated)
+    V30,
+    /// 40V (Deprecated)
+    V40,
+    /// 50V (Deprecated)
+    V50,
+}
+
+#[repr(C)]
+/// Charge Through Support
+#[derive(Debug, Clone, Copy, PartialEq, Eq, N)]
+pub enum Pd3p2ChargeThroughSupport {
+    /// the VPD does not support Charge Through
+    NotSupported = 0,
+    /// the VPD supports Charge Through
+    Supported,
+}
+
+#[repr(C)]
+/// VPD VDO. USB PD 3.2 VPD VDO (Section 6.4.4.3.1.9)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Pd3p2VpdVdo {
+    /// HW Version 0000b…1111b assigned by the VID owner
+    pub hw_version: u32,
```

Comment:
IMHO, using anything other than u32/i32 in Rust is a bit counterproductive unless the memory savings are really worth it. It does not have implicit conversions, so at some point the project becomes inundated with `from()/into()/try_from()/try_into()` and `as`.

In particular, for `try_from()/try_into()`, you now have to either consume the `Result` or propagate it upwards. This means returning `Result` from functions that otherwise could return nothing, thus the function signatures also become a bit more polluted for no reason.

 I don't think users will have thousands of this struct in use at the same time. 

WDYT?

Reply:
I gave it a try and I can just replace `u32` with `u8` without any other changes required. Note that [BitRead.read takes a Numeric trait](https://docs.rs/bitstream-io/latest/bitstream_io/read/trait.BitRead.html#tymethod.read). Reducing memory footprint also improves caching behavior and code size. And I do not really follow your arguments. Going from a small type to a big type is something that always works. The other way needs the Result. So you want to go with the smallest type possible, so that users can use `from()`/`into()` for the bigger types (when needed) instead of needing `try_from()/try_info()` for the smaller types.

Apart from that the crate currently only has functions to read the structures from encoded data. Once you want to do the reverse operation you need to validate that the structure actually encodes valid data.

So I only see disadvantages in storing more bits.